### PR TITLE
PWX-29422: Providing option to specify namespaces selector in migrati…

### PR DESCRIPTION
**Pull request by pp511**
…on spec.

**What type of PR is this?**
>feature

**What this PR does / why we need it**:
https://portworx.atlassian.net/browse/PWX-29422
Allow users to specify namespace selectors in migration spec.

**Does this PR change a user-facing CRD or CLI?**:
yes, 

**Is a release note needed?**:
Yes
* You can now provide namespace labels in MigrationSchedule spec to specify namespaces for migration.

**Does this change need to be cherry-picked to a release branch?**:
yes 23.5.0

**Notes**:
* I have added one function getMigrationNamespaces to get all unique namespaces for migration and referencing this list throughout our migration flow.
* 
 ```
MigrationSchedule ->
spec:
  template:
    spec:
        namespaceSelectors:
          ns: migrate

Namespace ->
kind: Namespace
metadata:
  labels:
    kubernetes.io/metadata.name: mysql
    ns: migrate
```
* Triggered  migration Jenkins job https://jenkins.pwx.dev.purestorage.com/job/Stork/view/Stork%2023.4-dev/job/23-4-dev-migration-k8s-1-24-0/463/ to make sure there are no regressions

### CODE REVIEW

File: pkg/migration/controllers/migration.go [hunk 1]
```diff
@@ -535,8 +543,44 @@ func (m *MigrationController) handle(ctx context.Context, migration *stork_api.M
...
+		label := key + "=" + val
+		namespaces, err := clientset.CoreV1().Namespaces().List(ctx, metav1.ListOptions{LabelSelector: label})
+		if err != nil {
+			return nil, err
+		}
+		for _, namespace := range namespaces.Items {
+			if _, ok := uniqueNamespaces[namespace.GetName()]; !ok {
+				migrationNamespaces = append(migrationNamespaces, namespace.GetName())
+			}
+			uniqueNamespaces[namespace.GetName()] = true
```
Comment (diptiranjanpx) [inline pkg/migration/controllers/migration.go:575]: Another approach could have been to only populate `uniqueNamespaces` map with namespaces and at the end create the list of keys and return that. The current approach is fine too.

**Review by diptiranjanpx — APPROVED** [review]

File: test/integration_test/migration_test.go [hunk 2]
```diff
@@ -456,6 +457,19 @@ func migrationLabelExcludeSelectorTest(t *testing.T) {
...
+	triggerMigrationTest(
+		t,
+		"ns-selector-test",
+		"cassandra",
+		[]string{"mysql-1-pvc"},
+		"namespace-selector-migration",
+		true,
+		true,
+		true,
+	)
```
Comment (adityadani) [inline test/integration_test/migration_test.go:470]: Lets add another app in a different namespace and ensure that that app is not migrated and cassandra gets migrated

File: pkg/migration/controllers/migration.go (hunk 1 above)
Comment (pp511) [reply pkg/migration/controllers/migration.go:569]: Updated as per suggestion.

File: test/integration_test/migration_test.go (hunk 2 above)
Comment (pp511) [reply test/integration_test/migration_test.go:501]: I have modified the test to run two apps in different namespace and only migrate cassandra for which namespace label matches.

File: pkg/apis/stork/v1alpha1/migration.go [hunk 3]
```diff
@@ -16,6 +16,7 @@ type MigrationSpec struct {
 	ClusterPair                  string            `json:"clusterPair"`
 	AdminClusterPair             string            `json:"adminClusterPair"`
 	Namespaces                   []string          `json:"namespaces"`
+	NamespaceSelectors           map[string]string `json:"namespaceSelectors"`
```
[the remark this answers is not in the document]
Comment (adityadani) [reply pkg/apis/stork/v1alpha1/migration.go:19]: You are right. My bad

**Review by adityadani — APPROVED** [review]

**Pull request merged**